oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,370 @@
+
+Sharing these dependencies has several important positive effects:
+
+- Enforces runtime consistency across all types of dependencies
+- Reduces runtime bundle size
+
+Note that "build time" and "dependency maintenance" are not mentioned above.
+
+Interestingly, because of the way Webpack module federation's dependency sharing is implemented, this may not have a large impact on build time or dependency maintenance. "Guest" MFEs still need to have the dependencies installed for the build to succeed, and artifacts for them will still be created and deployed - they just won't generally be used at runtime unless there's a version conflict.
+
+These dependencies still need to exist in the package.json file (and be installed by npm) for the build to succeed, and for IDE hinting, auto completion, linting, etc.
+
+Maintaining Dependency Consistency
+==================================
+
+This proposal fundamentally changes how we work with MFE dependencies, and will require us to adopt a more rigorous approach to ensuring dependency consistency and compatiblity.
+
+We expect that this may need to take a number of possible forms:
+
+Process
+-------
+
+We need to ensure maintainers and developers know what dependency versions to use, and when they need to upgrade to stay consistent. We also need a process to migrate through breaking changes in third-party dependencies.
+
+Best Practices
+--------------
+
+We need to ensure we minimize breaking changes in our own libraries (such as Paragon).
+
+Tooling
+-------
+
+Maintainers and developers should be warned of incompatibilities created by their PRs, or outside the repository by another project (such as the shell).
+
+Code Organization
+-----------------
+
+We could adopt a more drastic code organization scheme, such as a monorepo, which provides shared dependencies to core applications and libraries via a common package.json file, while otherwise maintaining applications' independent deployability.
+
+Alternately, we could create a new, versioned shared library which provides shared dependencies to MFE repositories as a single dependency (TBD how that would work).  This library could be paired with webpack aliases to prevent MFEs from needing to update their library import paths.
+
+One way or another, we're abandoning some of the independence of MFEs as originally envisioned years ago, in favor of more consistency between them, and we'll need a way to manage that.
+
+Out of Scope
+============
+
+There are various aspects of this which are considered out of scope for this OEP and its resulting reference implementation.
+
+- Opinions on which dependencies we should adopt going forward (such as redux or other state management solutions)
+- Implementation details of how module federation would be added in the frontend-plugin-framework.
+- Details of a monorepo approach.
+- How we will enforce dependency consistency.
+- How Tutor and other distributions will need to change to adopt module federation.
```

Comment:
Though not as critical as the above issue, I would most definitely appreciate some advance preparation for this item.

Reply:
I absolutely agree and have frantically trying to learn about Tutor so I can speak more intelligently about this. ;)  But for this OEP I think we do need to draw the line somewhere.

Reply:
I think this would be an ADR to this OEP